### PR TITLE
docs(training): add Lance-native dataloader migration design doc

### PR DESCRIPTION
**Pull request by ktinubu**
## Why

Epics in this repo require a committed design doc, and the Lance-native dataloader
epic (#1738) plus its four phases (#1739–#1742) cite this document by section.
This lands the read/dataloader-side design so the phase issues have a stable,
grounded reference. It is the sibling to the storage/write-side
`docs/design/lance-dataset-api-migration.md`.

## What changed

- Adds `docs/design/lance-dataloader-migration.md` (docs only; no code).
- Captures the decision to move the Lance read path from the HDF5-shaped shim to
  a map-style `LanceMapDataset` with a single projected `__getitems__` `take`, a
  pure `collate_fn`, R2 → node-local pre-staging, and removal of the hand-rolled
  samplers / `batch_size=None` / HDF5 path.
- Records the rejected alternatives (notably `SafeLanceDataset`) with
  pinned-source evidence, the DDP-correctness analysis, and the four-phase plan
  with its dependency graph.

## Test plan

Docs-only change — no runtime behavior. Verified by:

- `pre-commit` (mdformat, codespell, trailing-whitespace, EOF) green on the file.
- `/repo-review-full-no-comments` run to convergence: **0 BLOCK, 0 WARN** across
  code-health, comment-hygiene, and lance-review. Every cited repo line number
  and Lance/pylance 7.0.0 API claim was verified against installed source and
  live upstream docs (`SafeLanceDataset.__getitems__`, `LanceDataset` =
  `IterableDataset`, `ShardedBatchSampler` signature + k=256 reservoir,
  `to_numpy_ndarray()` on `FixedShapeTensorArray`, `data_storage_version="2.2"`).

Refs #1738

### CODE REVIEW

**Comment by ktinubu** [comment]
## Verification Results

Docs-only PR adding `docs/design/lance-dataloader-migration.md`. The `design/` tree is excluded from the mkdocs site (`mkdocs.yml` `exclude_docs`), so the verifiable promises are (a) the doc's internal links resolve and (b) its addition does not break the strict site build. Cross-reference accuracy (repo line numbers + pylance 7.0.0 API claims) was verified separately by `/repo-review-full-no-comments` to **0 BLOCK / 0 WARN** (sentinel `45433f8e`).

- [x] **Every relative link in the doc resolves to a real file** *(Level 2 — runs a real filesystem link-check; a failure means a dead in-repo link)*
  ```bash
  $ python3 - <<'PY'
  import re, pathlib
  doc = pathlib.Path("docs/design/lance-dataloader-migration.md")
  base = doc.parent; text = doc.read_text()
  bad=[]; checked=[]
  for href in re.findall(r'\[[^\]]+\]\(([^)]+)\)', text):
      if href.startswith(("http://","https://","#","mailto:")): continue
      target = href.split("#")[0]
      if not target: continue
      ok = (base/target).resolve().exists(); checked.append((href,ok))
      if not ok: bad.append(href)
  for h,ok in checked: print(f"  {'OK ' if ok else 'DEAD'}  {h}")
  print(f"checked={len(checked)} dead={len(bad)}"); assert not bad
  print("PASS")
  PY
    OK   lance-dataset-api-migration.md
    OK   lance-dataset-api-migration.md
    OK   lance-dataset-api-migration.md
  checked=3 dead=0
  PASS
  ```

- [x] **`mkdocs build --strict` succeeds with the doc present, and the doc stays excluded from the site** *(Level 1 — exercises the real site build end-to-end; strict mode fails on broken refs/nav)*
  ```bash
  $ uv run mkdocs build --strict 2>&1 | tail -4; echo "exit: ${PIPESTATUS[0]}"
  INFO    -  The following pages exist in the docs directory, but are not included in the "nav" configuration:
    - pr-readiness-loop.md
    - testing/mutmut.md
  INFO    -  Documentation built in 0.72 seconds
  exit: 0
  ```
  The new `design/lance-dataloader-migration.md` is **not** among the "not in nav" pages — confirming it is correctly excluded via `design/` and cannot silently break the strict build.

- [ ] **Cross-reference accuracy (repo L-numbers + Lance/pylance 7.0.0 API claims)** *(verified by `/repo-review-full-no-comments` round 5 → 0 BLOCK / 0 WARN against installed pylance source + live upstream docs; not re-run here as a standalone command)*

File: docs/design/lance-dataloader-migration.md [hunk 1]
```diff
@@ -0,0 +1,369 @@
...
+piece in the current datamodule — the `batch_size=None` dataloader, the
+batch-indexed `__getitem__`, and the `ShiftedBatchSampler` /
+`WithinChunkShuffledSampler` / `ShuffledSampler` family — exists to work around a
+single HDF5 limitation: **slow random access**. The workarounds keep reads
+*contiguous* so h5py stays fast, then claw back approximate shuffling on top of
+that contiguity.
+
+Lance's headline property is fast random access (its own docs market "100x
+faster random access" vs Parquet — see §10). So the premise those workarounds
+were built on no longer holds. This migration deletes the reason they exist
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply docs/design/lance-dataloader-migration.md]: Fixed in 13348acd: changed "its own docs market" → "its own docs claim" so the verb takes the quoted figure as its object.

File: docs/design/lance-dataloader-migration.md [hunk 2]
```diff
@@ -0,0 +1,369 @@
...
+node** (`prepare_data_per_node=True` by default), which is exactly what we want
+for node-local staging — provided `dataset_root` is a node-local path, not a
+shared network mount. Phase 3 verifies this.
+
+## §7 DDP analysis (load-bearing)
+
+The DDP risk in the *current* code has nothing to do with Lance. The training
+dataloader is `DataLoader(train_dataset, batch_size=None, sampler=ShiftedBatchSampler(...))`, the DDP config uses `strategy: ddp` on 4
+devices (`configs/trainer/ddp.yaml`), and nothing sets `use_distributed_sampler`
+→ Lightning's default `True`.
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply docs/design/lance-dataloader-migration.md:255]: Fixed in c3fd9df5: prefixed to `src/synth_setter/configs/trainer/ddp.yaml` (no repo-root `configs/` exists; now consistent with the other src-relative refs and doc-map.yaml).

File: docs/design/lance-dataloader-migration.md [hunk 3]
```diff
@@ -0,0 +1,369 @@
...
+- Cloudflare R2 zero egress (pre-staging is free): https://www.cloudflare.com/developer-platform/products/r2/
+
+**Internal (repo, grounded)**
+
+- `src/synth_setter/data/surge_datamodule.py` — `VSTDataset.__getitem__` per-column reads (L236–256); samplers (L275–390); `train_dataloader`/`batch_size=None` (L527–543); `prepare_data` (L462–472).
+- `src/synth_setter/data/lance_datamodule.py` — `LanceColumn.__getitem__` (L48–74); `LanceShardFile.live_dataset` getpid reopen (L108–122).
+- `src/synth_setter/pipeline/r2_io.py` — `r2_storage_options` (L188); `download_dir_no_overwrite`.
+- `src/synth_setter/pipeline/data/lance_shard.py` — `LANCE_DATA_STORAGE_VERSION = "2.2"` constant (L21), applied as the `data_storage_version` write kwarg (L111/L142).
+- `docs/design/lance-dataset-api-migration.md` — storage/write-API sibling; "training dataloader keeps rclone local-first" decision.
+- `configs/trainer/ddp.yaml` — `strategy: ddp`, `devices: 4`; no `use_distributed_sampler` override.
```
[the remark this answers is not in the document]
Comment (ktinubu) [reply docs/design/lance-dataloader-migration.md]: Fixed in c3fd9df5: prefixed to `src/synth_setter/configs/trainer/ddp.yaml` (no repo-root `configs/` exists; now consistent with the other src-relative refs and doc-map.yaml).

File: docs/design/lance-dataloader-migration.md [hunk 4]
```diff
@@ -0,0 +1,369 @@
...
+  `LANCE_DATA_STORAGE_VERSION = "2.2"` module constant
+  (`src/synth_setter/pipeline/data/lance_shard.py` L21) passed as the
+  `data_storage_version` write kwarg (L111/L142) — which has good random
+  access + projection. Rows are large: mel ≈ 400 KB, audio ≈ 1.4 MB,
+  music2latent ≈ 21 KB (from fake-mode shapes, `surge_datamodule.py`
+  `_get_fake_item`).
+
+**The two costs.** (1) N separate Lance scans per batch instead of one projected
+read. (2) Arrow → NumPy → Torch conversion per column. Both are artifacts of the
+h5-shaped interface, not of Lance.
```
Comment (khaledtin) [inline docs/design/lance-dataloader-migration.md]: **[synth-setter:block]** Incorrect grounding: the doc states shards are written with `data_storage_version="2.2"` and that `LANCE_DATA_STORAGE_VERSION = "2.2"`. The actual source constant is `LANCE_DATA_STORAGE_VERSION = "2.1"` (`src/synth_setter/pipeline/data/lance_shard.py:22`), and `"2.2"` appears nowhere in the repo — the source comment even notes `"2.1" equals the current default`. This is load-bearing: the whole §2 cost analysis (random-access + projection behavior) rests on the on-disk format version. Fix to `"2.1"` here and in the §10 References block.

File: docs/design/lance-dataloader-migration.md [hunk 5]
```diff
@@ -0,0 +1,369 @@
...
+(`torch/utils/data/_utils/fetch.py`), so Phase 2 pins it with a test that fails
+loudly if the `uv.lock`-pinned torch ever changes the collation path. The
+dataloader becomes boring on purpose:
+`DataLoader(ds, batch_size=B, shuffle=True, num_workers=n, collate_fn=collate, pin_memory=True)`.
+
+## §5 Tradeoffs considered
+
+### 5.1 Map-style vs iterable
+
+| Axis               | Map-style (chosen)                                 | Iterable (`lance.torch.data.LanceDataset`)                       |
```
Comment (khaledtin) [inline docs/design/lance-dataloader-migration.md:154]: **[ml-pipeline:block]** The headline dataloader sketch sets `shuffle=True` with `num_workers=n`, but the design never specifies a `worker_init_fn` or how the per-rank/per-epoch seed is set. Under DDP, Lightning's `DistributedSampler.set_epoch(epoch)` MUST be called every epoch or every epoch replays the identical permutation — a silent correctness bug of exactly the class §7 claims to eliminate. `set_epoch` is mentioned nowhere; the design fixes the sampler-replacement footgun but leaves the epoch-seeding obligation unspecified. Pin it explicitly and assert it in the Phase 2/§9 DDP test.

File: docs/design/lance-dataloader-migration.md [hunk 6]
```diff
@@ -0,0 +1,369 @@
...
+End state:
+
+```python
+class LanceMapDataset(torch.utils.data.Dataset):
+    def __init__(self, uri, columns, storage_options=None):
+        self.uri, self.columns = uri, columns
+        self._storage_options = storage_options  # persisted so the fork-time reopen keeps R2 creds
+        # Cached once; assumes an immutable staged split (no concurrent
+        # compaction/append), like LanceShardFile.
+        self._len = lance.dataset(uri, storage_options=storage_options).count_rows()
```
Comment (khaledtin) [inline docs/design/lance-dataloader-migration.md:121]: **[ml-pipeline:block]** The `LanceMapDataset` sketch is the design's contract object, but the only shape/dtype documentation for `__getitems__`/`__getitem__` lives in §4 prose, not as a method contract. Because the two index paths return different leading-dim shapes (the documented `collate_fn` foot-gun, lines 140–142), state the per-method shape/dtype contract on the methods themselves (e.g. `__getitems__` → `dict[str, ndarray]` of shape `(len(indices), *col_shape)`; `__getitem__` → single unwrapped sample) so the implementer cannot get it wrong.

**Review by khaledtin — CHANGES_REQUESTED** [review]
Multi-skill review of PR #1744 — 5 parallel passes (code-health, synth-setter-project-standards, comment-hygiene, ml-data-pipeline, lance-review). Each BLOCK below is posted as an individual unresolved inline thread; WARNs are collapsed into the `## Advisory (WARN) findings` section to keep the inline-thread list short. Findings on files outside the diff are anchored to the line in the diff that causes the staleness or rolled into the review body.

## Advisory (WARN) findings

- **[code-health:warn]** docs/design/lance-dataloader-migration.md:123 — The `LanceMapDataset` "end state" snippet uses compound tuple assignment (`self.uri, self.columns = uri, columns`) and a statement on the `def __len__` line; since Phase 2 implementers will copy this block, model the repo's one-statement-per-line / separate-line `return` style.
- **[code-health:warn]** docs/design/lance-dataloader-migration.md:131 — The snippet calls `self._live()` but `_live` is never defined, yet the design's §5.2/§7 correctness argument rests on its fork/`getpid`-reopen behavior. Inline a 3-line `_live()` stub (getpid guard + reopen threading `self._storage_options`) or point at the `LanceShardFile.live_dataset()` pattern it mirrors.
- **[code-health:warn]** docs/design/lance-dataloader-migration.md:127 — `__init__` opens `lance.dataset(...)` only for `count_rows()` then discards it (`self._ds = None`); a half-sentence clarifying the count-only open is intentionally discarded and the live handle is created lazily in `_live` would remove a leak-vs-intent ambiguity.
- **[synth-setter:warn]** docs/design/lance-dataloader-migration.md:79 — Wrong line citation: the constant is at `lance_shard.py` L22, not L21 (L21 is the comment above it); the same mis-cite repeats in the References block. Write-kwarg lines are 112/143, not 111/142 (within `~` tolerance).
- **[synth-setter:warn]** docs/design/lance-dataloader-migration.md:372 — References-section restates the same `"2.2"`/L21 errors; correct to `"2.1"`, `L22`, `L112/L143` here too.
- **[synth-setter:warn]** docs/design/lance-dataloader-migration.md:124 — Credential note for implementers: persisting `storage_options` (plaintext R2 `access_key_id`/`secret_access_key` from `r2_storage_options()`) as a long-lived instance attribute replicated into every worker is only needed on the streaming fallback; for the default local path (§6) it should be `None`/absent. Add one sentence so Phase 2 doesn't wire creds in unconditionally (also: ml-pipeline).
- **[comment-hygiene:warn]** docs/design/lance-dataloader-migration.md:128 — Redundant inline comments: L128 `# reopened after fork (getpid guard in _live)` and L132 `# reopen on fork (reuse existing pattern)` state the same fact; keep L128 and tighten the call-site to add the local fact (e.g. `# getpid-guarded handle`) (also: code-health).
- **[ml-pipeline:warn]** docs/design/lance-dataloader-migration.md:122 — `columns`/`storage_options` enter via constructor args but the doc never states they originate from Hydra/YAML config (config-externalized); say the projection list and R2 options come from the datamodule config so an implementer doesn't inline a literal column list.
- **[ml-pipeline:warn]** docs/design/lance-dataloader-migration.md:148 — The design leans on a torch-internal (`_MapDatasetFetcher` passing the `__getitems__` dict straight to `collate_fn`) and pins it with a test; state the failure mode (shape corruption / double-stacking vs. exception) and assertion target (collate receives a dict, not `list[dict]`) so the pinning test actually catches a regression.
- **[ml-pipeline:warn]** docs/design/lance-dataloader-migration.md:336 — The read-equivalence test asserts value equality but never binds "projection minimality" to an observable (only `self.columns` scanned). Since §5.2 makes no-over-read the core reason to reject `SafeLanceDataset`, add a test that fails when an unprojected `take` sneaks in, else the central perf claim stays unverified.
- **[ml-pipeline:warn]** docs/design/lance-dataloader-migration.md:134 — `combine_chunks().to_numpy_ndarray()` is asserted for `FixedShapeTensorArray` columns but no contract/fallback is given for non-fixed-shape columns (variable-length `audio`, scalar `param_array`); state which column types the path assumes or the implementer hits a conversion error on the first non-tensor column.
- **[lance:warn]** docs/design/lance-dataloader-migration.md:357 — §10 cites the Lance repo as `github.com/lance-format/lance` (resolves, currently-canonical after a recent rename), but most cross-references still use `lancedb/lance`; `main`-branch deep links to `data.py`/`sampler.py` will drift — pin to a commit-SHA permalink.
- **[lance:warn]** docs/design/lance-dataloader-migration.md:356 — Could not independently confirm the `cache=True` kwarg name (line 248) against a live signature within the lookup budget; `IterableDataset`, `SafeLanceDataset`, and `batch_readahead=8` all verified verbatim. Quick grounding check on the exact `cache` kwarg before merge.

File: docs/design/lance-dataloader-migration.md [hunk 7]
```diff
@@ -0,0 +1,369 @@
...
+Under that default, Lightning **rewrites the dataloader's sampler** for DDP. Its
+behavior is version-dependent and is exactly the thing we should not rely on
+silently:
+
+- Lightning special-cases plain `SequentialSampler` / `RandomSampler` and
+  cleanly swaps in a `DistributedSampler`; for *custom* samplers it instead
+  wraps/replaces with a warning ("Replacing custom sampler with distributed
+  version …"). See §10 (Lightning trainer docs + issue #5271).
+- With a custom `ShiftedBatchSampler` + `batch_size=None`, the most likely
+  outcome is correct *sharding* (each integer index → one contiguous batch via
```
Comment (khaledtin) [inline docs/design/lance-dataloader-migration.md:266]: Consistency nit in the load-bearing §7. This says the likely DDP outcome is "correct *sharding* (each **integer index** → one contiguous batch via `__getitem__`)", but §2 (and the actual code) establish that `ShiftedBatchSampler` is a `BatchSampler` subclass whose `__iter__` yields `(start, stop)` **tuples** — `__getitem__` receives a tuple and slices, there is no integer index on the training path. Since this section's whole point is the precise sampler-rewrite mechanism, the imprecise "integer index" framing undercuts it. Suggest: "each `(start, stop)` index → one contiguous batch via `__getitem__`". Verified: `surge_datamodule.py:362` (`class ShiftedBatchSampler(torch.utils.data.BatchSampler)`) and `:388` (`yield (i * self.batch_size + offset, ...)`).

File: docs/design/lance-dataloader-migration.md [hunk 8]
```diff
@@ -0,0 +1,369 @@
+# Design Doc: Lance-Native Dataloader Migration
+
+> **Status**: Draft
```
Comment (khaledtin) [inline docs/design/lance-dataloader-migration.md:3]: Status mismatch (doc-lifecycle, holistic). The header marks this **Draft**, but the doc is being committed specifically as the *stable grounding reference* that four already-filed phase issues (#1739–#1742, all OPEN and citing this doc by section) point at, and §4 records a finalized **Decision** with rejected alternatives pinned. A reader arriving from a phase issue sees "Draft" and reasonably distrusts the decision as still-in-flux. Recommend promoting to `Accepted`/`Approved` on merge (or note explicitly that the decision is accepted while wording is draft) so the committed status matches its load-bearing role.

**Review by khaledtin — COMMENTED** [review]
Personal overall doc review (holistic pass, separate from the multi-skill review). Two additional findings; both verified against the tree. Grounding is otherwise strong — sampler names (`ShiftedBatchSampler`/`WithinChunkShuffledSampler`/`ShuffledSampler`), the `(start, stop)` tuple yield, and all five cited issues (#1738–#1742, matching titles) check out.

**Comment by ktinubu** [comment]
Closing after stale-PR triage: superseded by the as-built Lance loader implementation in #1783 and #1834, followed by legacy-loader removal in #2075.
